sig-scalability/blogs/k8s-services-scalability-issues.md

```diff
@@ -0,0 +1,133 @@
+# Known Scalability Issues with Kubernetes Services
+
+_by Shyam JVS, Google Inc (with inputs from Brian Grant, Wojciech Tyczynski & Dan Winship)_
+
+**June 2018**
+
+This document serves as a catalog of issues we've known/discovered with kubernetes services as of June 2018, focusing on their scalability/performance. The purpose of the document is to make the information common knowledge for the community, so we can work together towards improving it. Listing them below in no particular order.
+
+
+## Iptables packet processing performance when large number of service ports exist
+
+#### Issue
+
+Iptables can be slow in packet processing when a large number of services exist. As the number of service ports increases, the KUBE-SERVICES chain gets longer and since it’s also evaluated very frequently that can bring down the performance. There have been some recent improvements in this area ([#56164](https://github.com/kubernetes/kubernetes/pull/56164), [#57461](https://github.com/kubernetes/kubernetes/pull/57461), [#60306](https://github.com/kubernetes/kubernetes/pull/60306)) - but we still need to measure and come up with safe bounds for number of services that can be 'decently' supported (10k seems to be a reasonable estimate from [past discussions](https://github.com/kubernetes/kubernetes/issues/48938#issue-243000172)). Also, further improvements might be required based on those results.
```

Comment:
We could turn the linear search into a binary search pretty easily, we have never had a driving demand for it.  IPVS and other modes seem to be better in almost every way, so we probably won't.